From cf9195ef74c689b09ae7874b16e44435cb7fe71b Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Wed, 24 Aug 2016 09:47:52 -0700
Subject: [PATCH] DOCKER-918: Allow docker build to copy files to a directory
 when no trailing slash is used Reviewed by: Orlando Vazquez
 <orlando@joyent.com>

---
 lib/build.js       | 24 +++++++++++++++++++++++-
 test/test_build.js | 26 ++++++++++++++++++++++++++
 2 files changed, 49 insertions(+), 1 deletion(-)

diff --git a/lib/build.js b/lib/build.js
index a2ce9d7..ae954a8 100644
--- a/lib/build.js
+++ b/lib/build.js
@@ -1528,6 +1528,28 @@ Builder.prototype.getCopyInfo = function getCopyInfo(cmd, opts, callback) {
         return;
     }
 
+    // There is one docker edge case when a file is being copied to a directory,
+    // but dest has no trailing slash, the file gets copied into the directory
+    // as /dest/file. So if dest is a directory here, we add a trailing slash.
+    if (filepaths.length === 1 && dest[dest.length - 1] !== '/') {
+        var globalDest = path.join(builder.containerRootDir, resolvedDest);
+        var destStat = null;
+        try {
+            destStat = fs.statSync(globalDest);
+        } catch (e) {
+            if (e.code !== 'ENOENT') {
+                callback(new Error(
+                    util.format('Unable to stat path %s, error: %s', dest, e)));
+                return;
+            }
+            // No such file/dir, then all is good.
+        }
+        if (destStat && destStat.isDirectory()) {
+            builder.log.debug('dest is a directory, adding a trailing slash');
+            resolvedDest += '/';
+        }
+    }
+
     var calcOpts = {
         origPath: '', // Updated in mapFn
         destPath: resolvedDest,
@@ -1556,7 +1578,7 @@ Builder.prototype.getCopyInfo = function getCopyInfo(cmd, opts, callback) {
             return;
         }
 
-        if (copyInfos.length > 1 && dest[dest.length - 1] != '/') {
+        if (copyInfos.length > 1 && dest[dest.length - 1] !== '/') {
             callback(new Error(util.format(
                 'When using %s with more than one source '
                 + 'file, the destination must be a '
diff --git a/test/test_build.js b/test/test_build.js
index b5e37be..683fa3a 100644
--- a/test/test_build.js
+++ b/test/test_build.js
@@ -781,6 +781,32 @@ tape('addMissingFile', function (t) {
 });
 
 
+// DOCKER-918: Test copying a file to a directory, but leave out the trailing
+// slash. Following docker/docker this should be allowed.
+tape('copy to dir without trailing slash', function (t) {
+    var fileAndContents = {
+        'Dockerfile': [
+            'FROM busybox',
+            'RUN mkdir /adir',
+            'COPY file.txt /adir'
+        ].join('\n'),
+        'file.txt': 'hello'
+    };
+    testBuildContents(t, fileAndContents, function (err, result) {
+        var builder = result.builder;
+        if (showError(t, err, builder)) {
+            return;
+        }
+
+        var root = builder.containerRootDir;
+        var buf = fs.readFileSync(path.join(root, '/adir/file.txt'));
+        t.equal(buf.toString(), fileAndContents['file.txt'], 'Check file.txt');
+
+        testEnd(t, builder);
+    });
+});
+
+
 tape('workdir', function (t) {
     var fileAndContents = {
         'Dockerfile': [
-- 
2.21.0

